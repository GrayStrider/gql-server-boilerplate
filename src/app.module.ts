import { Module } from '@nestjs/common'
import { AppController } from './app.controller'
import { AppService } from './app.service'

@Module({
	imports: [],
	controllers: [AppController],
	providers: [AppService]
})
// eslint-disable-next-line @typescript-eslint/no-extraneous-class
export class AppModule {}
