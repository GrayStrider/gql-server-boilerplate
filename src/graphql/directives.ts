import {GraphQLEnumValue, GraphQLField} from 'graphql'
import {SchemaDirectiveVisitor} from 'graphql-tools'

export default class DeprecatedDirective extends SchemaDirectiveVisitor {
	
	public visitFieldDefinition (field: GraphQLField<unknown, unknown>) {
		
		field.isDeprecated = true
		field.deprecationReason = this.args.reason
		
	}
	
	public visitEnumValue (value: GraphQLEnumValue) {
		
		value.isDeprecated = true
		value.deprecationReason = this.args.reason
		
	}
	
}
