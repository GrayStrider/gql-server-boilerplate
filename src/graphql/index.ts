import genericApolloServer from '@/graphql/apollo/genericServer'
import uploadSchema from '@/graphql/schemas/uploadSchema'
import context, {Context} from '@/graphql/apollo/context'
import dataSources from '@/graphql/apollo/datasources'
import DeprecatedDirective from '@/graphql/directives'
import plainSchema from '@/graphql/schemas/plainSchema'
import createSchema from '@/graphql/type-graphql/createSchema'
import formatError from '@/graphql/apollo/formatError'

export {
	genericApolloServer,
	uploadSchema,
	Context,
	context,
	dataSources,
	formatError,
	DeprecatedDirective,
	plainSchema,
	createSchema
}
