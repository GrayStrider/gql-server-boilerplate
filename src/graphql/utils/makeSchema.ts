import {IExecutableSchemaDefinition, makeExecutableSchema} from 'graphql-tools'
import {importSchema} from 'graphql-import'

export default async function makeSchema<TContext> (schemaPath: string, params: Omit<IExecutableSchemaDefinition, 'typeDefs'>) {
	
	const typeDefs = await importSchema(schemaPath, {})
	return makeExecutableSchema<TContext>({...params, typeDefs})
	
}

