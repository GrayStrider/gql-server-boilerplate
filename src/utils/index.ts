import signale from 'signale'
import * as RD from 'ramda-adjunct'
import {Promise as bb} from 'bluebird'
import gql from 'graphql-tag'
import Errors from '@/utils/Errors'
import log from '@/utils/log'
import flattenGQLResponse from '@/utils/flattenGQLResponse'
import warn from '@/utils/warn'
import sleep from '@/utils/sleep'
import consoleWrite from '@/utils/consoleWrite'

signale.config({
	displayScope: true,
	displayBadge: false,
	displayDate: false,
	displayFilename: true,
	displayLabel: true,
	displayTimestamp: true,
	underlineLabel: true,
	underlineMessage: false,
	underlinePrefix: false,
	underlineSuffix: false,
	uppercaseLabel: false,
})

export * from '@/utils/Errors'
export * from '@/utils/testing/supertest'

export {
	Errors,
	RD,
	signale,
	log,
	bb,
	flattenGQLResponse,
	warn,
	sleep,
	consoleWrite,
	gql
}
