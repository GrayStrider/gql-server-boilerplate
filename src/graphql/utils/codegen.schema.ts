import {promises} from 'fs'
import {filter, test as regexpTest, isEmpty} from 'ramda'
import {buildSchema, GraphQLSchema} from 'graphql'
import {mergeSchemas} from 'graphql-tools'
import {bb} from '@/utils'

const {readdir, readFile} = promises

const PATH = 'src/graphql/generated/'

async function main (): Promise<GraphQLSchema> {
	
	const files = await readdir(PATH)
		.then(filter(regexpTest(/.graphql$/u)))
	
	if (isEmpty(files)) {

		const defaultSchema = `
        type Query {
            placeholder: String
        }
        
        type Mutation {
        		placeholder: String
        }
		`
		return buildSchema(defaultSchema)
	
	}
	
	const schemas = await bb.map(files,
		async fileName =>
			readFile(PATH + fileName, 'utf8'))
	
	return mergeSchemas({schemas})
	
}


export default main()
	.then(res => res)
