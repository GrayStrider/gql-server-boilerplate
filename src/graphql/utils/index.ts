import makeSchema from '@/graphql/utils/makeSchema'

export {
	makeSchema,
}
