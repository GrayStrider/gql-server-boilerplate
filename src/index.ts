import main from '@/server'
import {signale} from '@/utils'
import {NODE_ENV, PORT, HOST} from '@config'

process.on('uncaughtException', error => {
	
	console.error(`uncaught Exception: ${error.message}`)
	console.error(error)
	// eslint-disable-next-line no-process-exit
	process.exit(1)
	
})

process.on('unhandledRejection', error => {
	
	console.error(error)
	
})

main()
	.then(app => app
		.listen(PORT, () =>
			signale.success(`Server started at http://${HOST}:${PORT}`)))
	.catch(err => {
	
		signale.error('Error in main:')
		console.error(err)
	
		// eslint-disable-next-line no-process-exit
		if (NODE_ENV === 'production' || NODE_ENV === 'test') process.exit(1)
		throw err
	
	})
