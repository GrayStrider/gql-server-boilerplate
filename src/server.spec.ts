import {ASTNode} from 'graphql'
import main from 'src/server'
import {sleep, supertest, Test, SuperTest} from 'src/utils'

describe('server', () => {
	
	let request: SuperTest<Test>
	let post: <T>(query: ASTNode) => Promise<T>
	
	beforeAll(async () => {
		
		({request, post} = supertest(await main(), '/example'))
		
	})
	
	afterAll(async () => {
		
		// Avoid jest open handle error
		await sleep(500)
		
	})
	
	it('should start the server and redirect', async () => {
		
		expect.assertions(1)
		const res = await request.get('/foobar')
		expect(res.status).toBe(302)
		
	})

})
