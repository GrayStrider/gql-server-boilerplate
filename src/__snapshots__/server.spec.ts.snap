// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`server should return paginated response 1`] = `
Object {
  "hasMore": false,
  "items": Array [
    Object {
      "id": "f33e2858-f294-4c6d-91ae-7f78a3efca1b",
    },
  ],
  "total": 1,
}
`;
